skimage/segmentation/random_walker_segmentation.py

```diff
@@ -208,16 +208,19 @@ def _solve_linear_system(lap_sparse, B, tol, mode):
                 )
             M = None
         elif mode == 'cg_j':
-            M = sparse.diags(1.0 / lap_sparse.diagonal())
+            n = lap_sparse.shape[-1]
+            M = sparse.dia_array((1.0 / lap_sparse.diagonal(), 0), shape=(n, n))
         else:
             # mode == 'cg_mg'
             lap_sparse = lap_sparse.tocsr()
+            lap_sparse.indptr = lap_sparse.indptr.astype('int32')
+            lap_sparse.indices = lap_sparse.indices.astype('int32')
```

Comment:
I don't know what size `lap_sparse` can get to, but in general it seems prudent to check before downcasting. Assuming no restrictions on `lap_sparse`, we could check it the way `scipy.sparse.linalg`and `csgraph` do. That is, something like:
```suggestion
            # check for safe downcasting to int32
            max_value = np.iinfo('int32').max

            if lap_sparse.indptr[-1] > max_value:  # indptr[-1] is max b/c indptr always sorted
                raise ValueError("indptr values too large for pyamg")
            if max(*lap_sparse.shape) > max_value:  # only check large enough arrays
                if np.any(lap_sparse.indices > max_value):
                    raise ValueError("indices values too large for pyamg")

            lap_sparse.indptr = lap_sparse.indptr.astype('int32', copy=False)
            lap_sparse.indices = lap_sparse.indices.astype('int32', copy=False)
```
If this occurs frequently we might want to create a function to downcast or raise.

Reply:
I made a PR onto @jarrodmillman 's branch to wrap this as a function so it could be used elsewhere.
I don't see other places for pyamg.

But in pixel_graph we still use `csr_matrix` until `csgraph` support csr_array.  Converting graph to int32 indices seems to work with csgraph. That is, all your skimage tests pass if I use the `_safe_downcast_indices` function to downcast before calling csgraph.
I made this change as a second commit in that PR to Jarrod's branch. I don't know how to test it beyond the tests already written here.

Do you want me to move the safe function to utils or somewhere else?